src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs

```diff
@@ -3018,5 +3018,524 @@ void F(object o)
+  IL_000f:  ldloc.2
+  IL_0010:  callvirt   ""int string.Length.get""
+  IL_0015:  stloc.1
+  // sequence point: <hidden>
+  IL_0016:  ldloc.1
+  IL_0017:  ldc.i4.2
+  IL_0018:  bgt.s      IL_0031
+  IL_001a:  br.s       IL_0062
+  IL_001c:  ldloc.2
+  IL_001d:  brfalse.s  IL_0064
+  IL_001f:  ldloc.2
+  IL_0020:  callvirt   ""int string.Length.get""
+  IL_0025:  stloc.1
+  // sequence point: <hidden>
+  IL_0026:  ldloc.1
+  IL_0027:  ldc.i4.2
+  IL_0028:  bgt.s      IL_0035
+  IL_002a:  br.s       IL_0064
+  // sequence point: when b1
+  IL_002c:  ldarg.1
+  IL_002d:  brfalse.s  IL_001c
+  // sequence point: throw null;
+  IL_002f:  ldnull
+  IL_0030:  throw
+  // sequence point: <hidden>
+  IL_0031:  ldc.i4.0
+  IL_0032:  stloc.0
+  IL_0033:  br.s       IL_0037
+  IL_0035:  ldc.i4.2
+  IL_0036:  stloc.0
+  // sequence point: when b2
+  IL_0037:  ldarg.2
+  IL_0038:  brfalse.s  IL_0041
+  IL_003a:  ldloc.0
+  IL_003b:  brfalse.s  IL_0048
+  IL_003d:  ldloc.0
+  IL_003e:  ldc.i4.2
+  IL_003f:  beq.s      IL_0048
+  // sequence point: <hidden>
+  IL_0041:  ldloc.0
+  IL_0042:  brfalse.s  IL_0062
+  IL_0044:  ldloc.0
+  IL_0045:  ldc.i4.2
+  IL_0046:  beq.s      IL_0064
+  // sequence point: System.Console.WriteLine((s, b1, b2, j.ToString()));
+  IL_0048:  ldarg.0
+  IL_0049:  ldarg.1
+  IL_004a:  ldarg.2
+  IL_004b:  ldloca.s   V_1
+  IL_004d:  call       ""string int.ToString()""
+  IL_0052:  newobj     ""System.ValueTuple<string, bool, bool, string>..ctor(string, bool, bool, string)""
+  IL_0057:  box        ""System.ValueTuple<string, bool, bool, string>""
+  IL_005c:  call       ""void System.Console.WriteLine(object)""
+  // sequence point: break;
+  IL_0061:  ret
+  // sequence point: throw null;
+  IL_0062:  ldnull
+  IL_0063:  throw
+  // sequence point: throw null;
+  IL_0064:  ldnull
+  IL_0065:  throw
+}
+", source: source, sequencePoints: "C.M");
+        }
+
+        [Fact, WorkItem(55668, "https://github.com/dotnet/roslyn/issues/55668")]
+        public void SharedWhenExpression_Multiples()
+        {
+            // The `b3` condition ends up in the `when` clause on four leaves in the DAG
+            // and `b1` ends up in two leaves
+            var source = @"
+int count = 0;
+
+foreach (int i1 in new[] { 0, 1 })
+    foreach (int i2 in new[] { 0, 1 })
+        foreach (int i3 in new[] { 0, 1 })
+            foreach (bool b0 in new[] { false, true })
+                foreach (bool b1 in new[] { false, true })
+                    foreach (bool b2 in new[] { false, true })
+                        foreach (bool b3 in new[] { false, true })
+                        {
+                            count++;
+                            if (M(i1, i2, i3, b0, b1, b2, b3) != M2(i1, i2, i3, b0, b1, b2, b3))
+                                throw null;
+                        }
+
+System.Console.Write(count);
+
+static string M(int i1, int i2, int i3, bool b0, bool b1, bool b2, bool b3)
+{
+    object o = null;
+    switch (i1, i2, i3)
```

Comment:
:memo: Below is the dump of this DAG. Notice that states 10, 14, 18 and 20 will match when `b3` is true, and states 8 and 17 match when `b1` is true.

<details><summary>DAG dump</summary>
```
State 0
    1. [case (var x, var y, var z) when f(x, y, z):] AND(t1=DagFieldEvaluation(t0.i1), t2=DagFieldEvaluation(t0.i2), t3=DagFieldEvaluation(t0.i3)) BIND[x=t1; y=t2; z=t3] WHEN[f(x, y, z)]
    2. [case (not 0, 0, _) when b0:] AND(t1=DagFieldEvaluation(t0.i1), Not (?DagValueTest(t1 == ConstantValueDefault(0: Int32))), t2=DagFieldEvaluation(t0.i2), ?DagValueTest(t2 == ConstantValueDefault(0: Int32))) WHEN[b0]
    3. [case (_, not 0, 0) when b1:] AND(t2=DagFieldEvaluation(t0.i2), Not (?DagValueTest(t2 == ConstantValueDefault(0: Int32))), t3=DagFieldEvaluation(t0.i3), ?DagValueTest(t3 == ConstantValueDefault(0: Int32))) WHEN[b1]
    4. [case (0, _, not 0) when b2:] AND(t1=DagFieldEvaluation(t0.i1), ?DagValueTest(t1 == ConstantValueDefault(0: Int32)), t3=DagFieldEvaluation(t0.i3), Not (?DagValueTest(t3 == ConstantValueDefault(0: Int32)))) WHEN[b2]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] AND(t1=DagFieldEvaluation(t0.i1), ?DagValueTest(t1 == ConstantValueDefault(0: Int32)))
    7. [case (_, 0, _):] AND(t2=DagFieldEvaluation(t0.i2), ?DagValueTest(t2 == ConstantValueDefault(0: Int32)))
    8. [case (_, _, 0):] AND(t3=DagFieldEvaluation(t0.i3), ?DagValueTest(t3 == ConstantValueDefault(0: Int32)))
    Test: t1=DagFieldEvaluation(t0.i1)
    TrueBranch: 1
State 1
    1. [case (var x, var y, var z) when f(x, y, z):] AND(t2=DagFieldEvaluation(t0.i2), t3=DagFieldEvaluation(t0.i3)) BIND[x=t1; y=t2; z=t3] WHEN[f(x, y, z)]
    2. [case (not 0, 0, _) when b0:] AND(Not (?DagValueTest(t1 == ConstantValueDefault(0: Int32))), t2=DagFieldEvaluation(t0.i2), ?DagValueTest(t2 == ConstantValueDefault(0: Int32))) WHEN[b0]
    3. [case (_, not 0, 0) when b1:] AND(t2=DagFieldEvaluation(t0.i2), Not (?DagValueTest(t2 == ConstantValueDefault(0: Int32))), t3=DagFieldEvaluation(t0.i3), ?DagValueTest(t3 == ConstantValueDefault(0: Int32))) WHEN[b1]
    4. [case (0, _, not 0) when b2:] AND(?DagValueTest(t1 == ConstantValueDefault(0: Int32)), t3=DagFieldEvaluation(t0.i3), Not (?DagValueTest(t3 == ConstantValueDefault(0: Int32)))) WHEN[b2]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] ?DagValueTest(t1 == ConstantValueDefault(0: Int32))
    7. [case (_, 0, _):] AND(t2=DagFieldEvaluation(t0.i2), ?DagValueTest(t2 == ConstantValueDefault(0: Int32)))
    8. [case (_, _, 0):] AND(t3=DagFieldEvaluation(t0.i3), ?DagValueTest(t3 == ConstantValueDefault(0: Int32)))
    Test: t2=DagFieldEvaluation(t0.i2)
    TrueBranch: 2
State 2
    1. [case (var x, var y, var z) when f(x, y, z):] t3=DagFieldEvaluation(t0.i3) BIND[x=t1; y=t2; z=t3] WHEN[f(x, y, z)]
    2. [case (not 0, 0, _) when b0:] AND(Not (?DagValueTest(t1 == ConstantValueDefault(0: Int32))), ?DagValueTest(t2 == ConstantValueDefault(0: Int32))) WHEN[b0]
    3. [case (_, not 0, 0) when b1:] AND(Not (?DagValueTest(t2 == ConstantValueDefault(0: Int32))), t3=DagFieldEvaluation(t0.i3), ?DagValueTest(t3 == ConstantValueDefault(0: Int32))) WHEN[b1]
    4. [case (0, _, not 0) when b2:] AND(?DagValueTest(t1 == ConstantValueDefault(0: Int32)), t3=DagFieldEvaluation(t0.i3), Not (?DagValueTest(t3 == ConstantValueDefault(0: Int32)))) WHEN[b2]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] ?DagValueTest(t1 == ConstantValueDefault(0: Int32))
    7. [case (_, 0, _):] ?DagValueTest(t2 == ConstantValueDefault(0: Int32))
    8. [case (_, _, 0):] AND(t3=DagFieldEvaluation(t0.i3), ?DagValueTest(t3 == ConstantValueDefault(0: Int32)))
    Test: t3=DagFieldEvaluation(t0.i3)
    TrueBranch: 3
*State 3
    1. [case (var x, var y, var z) when f(x, y, z):] MATCH BIND[x=t1; y=t2; z=t3] WHEN[f(x, y, z)]
    2. [case (not 0, 0, _) when b0:] AND(Not (?DagValueTest(t1 == ConstantValueDefault(0: Int32))), ?DagValueTest(t2 == ConstantValueDefault(0: Int32))) WHEN[b0]
    3. [case (_, not 0, 0) when b1:] AND(Not (?DagValueTest(t2 == ConstantValueDefault(0: Int32))), ?DagValueTest(t3 == ConstantValueDefault(0: Int32))) WHEN[b1]
    4. [case (0, _, not 0) when b2:] AND(?DagValueTest(t1 == ConstantValueDefault(0: Int32)), Not (?DagValueTest(t3 == ConstantValueDefault(0: Int32)))) WHEN[b2]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] ?DagValueTest(t1 == ConstantValueDefault(0: Int32))
    7. [case (_, 0, _):] ?DagValueTest(t2 == ConstantValueDefault(0: Int32))
    8. [case (_, _, 0):] ?DagValueTest(t3 == ConstantValueDefault(0: Int32))
    FalseBranch: 4
State 4
    2. [case (not 0, 0, _) when b0:] AND(Not (?DagValueTest(t1 == ConstantValueDefault(0: Int32))), ?DagValueTest(t2 == ConstantValueDefault(0: Int32))) WHEN[b0]
    3. [case (_, not 0, 0) when b1:] AND(Not (?DagValueTest(t2 == ConstantValueDefault(0: Int32))), ?DagValueTest(t3 == ConstantValueDefault(0: Int32))) WHEN[b1]
    4. [case (0, _, not 0) when b2:] AND(?DagValueTest(t1 == ConstantValueDefault(0: Int32)), Not (?DagValueTest(t3 == ConstantValueDefault(0: Int32)))) WHEN[b2]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] ?DagValueTest(t1 == ConstantValueDefault(0: Int32))
    7. [case (_, 0, _):] ?DagValueTest(t2 == ConstantValueDefault(0: Int32))
    8. [case (_, _, 0):] ?DagValueTest(t3 == ConstantValueDefault(0: Int32))
    Test: ?DagValueTest(t1 == ConstantValueDefault(0: Int32))
    TrueBranch: 5
    FalseBranch: 12
State 5 REMAINING t1:[0..0]
    3. [case (_, not 0, 0) when b1:] AND(Not (?DagValueTest(t2 == ConstantValueDefault(0: Int32))), ?DagValueTest(t3 == ConstantValueDefault(0: Int32))) WHEN[b1]
    4. [case (0, _, not 0) when b2:] Not (?DagValueTest(t3 == ConstantValueDefault(0: Int32))) WHEN[b2]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] MATCH
    Test: ?DagValueTest(t2 == ConstantValueDefault(0: Int32))
    TrueBranch: 6
    FalseBranch: 7
State 6 REMAINING t1:[0..0] t2:[0..0]
    4. [case (0, _, not 0) when b2:] Not (?DagValueTest(t3 == ConstantValueDefault(0: Int32))) WHEN[b2]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] MATCH
    Test: ?DagValueTest(t3 == ConstantValueDefault(0: Int32))
    TrueBranch: 10
    FalseBranch: 9
State 7 REMAINING t1:[0..0] t2:[-2147483648..-1],[1..2147483647]
    3. [case (_, not 0, 0) when b1:] ?DagValueTest(t3 == ConstantValueDefault(0: Int32)) WHEN[b1]
    4. [case (0, _, not 0) when b2:] Not (?DagValueTest(t3 == ConstantValueDefault(0: Int32))) WHEN[b2]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] MATCH
    Test: ?DagValueTest(t3 == ConstantValueDefault(0: Int32))
    TrueBranch: 8
    FalseBranch: 9
*State 8 REMAINING t3:[0..0] t1:[0..0] t2:[-2147483648..-1],[1..2147483647]
    3. [case (_, not 0, 0) when b1:] MATCH WHEN[b1]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] MATCH
    FalseBranch: 10
*State 9 REMAINING t3:[-2147483648..-1],[1..2147483647] t1:[0..0] t2:[-2147483648..2147483647]
    4. [case (0, _, not 0) when b2:] MATCH WHEN[b2]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] MATCH
    FalseBranch: 10
*State 10 REMAINING t3:[-2147483648..2147483647] t1:[0..0] t2:[-2147483648..2147483647]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    6. [case (0, _, _):] MATCH
    FalseBranch: 11
*State 11 REMAINING t3:[-2147483648..2147483647] t1:[0..0] t2:[-2147483648..2147483647]
    6. [case (0, _, _):] MATCH
State 12 REMAINING t1:[-2147483648..-1],[1..2147483647]
    2. [case (not 0, 0, _) when b0:] ?DagValueTest(t2 == ConstantValueDefault(0: Int32)) WHEN[b0]
    3. [case (_, not 0, 0) when b1:] AND(Not (?DagValueTest(t2 == ConstantValueDefault(0: Int32))), ?DagValueTest(t3 == ConstantValueDefault(0: Int32))) WHEN[b1]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    7. [case (_, 0, _):] ?DagValueTest(t2 == ConstantValueDefault(0: Int32))
    8. [case (_, _, 0):] ?DagValueTest(t3 == ConstantValueDefault(0: Int32))
    Test: ?DagValueTest(t2 == ConstantValueDefault(0: Int32))
    TrueBranch: 13
    FalseBranch: 16
*State 13 REMAINING t1:[-2147483648..-1],[1..2147483647] t2:[0..0]
    2. [case (not 0, 0, _) when b0:] MATCH WHEN[b0]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    7. [case (_, 0, _):] MATCH
    FalseBranch: 14
*State 14 REMAINING t1:[-2147483648..-1],[1..2147483647] t2:[0..0]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    7. [case (_, 0, _):] MATCH
    FalseBranch: 15
*State 15 REMAINING t1:[-2147483648..-1],[1..2147483647] t2:[0..0]
    7. [case (_, 0, _):] MATCH
State 16 REMAINING t1:[-2147483648..-1],[1..2147483647] t2:[-2147483648..-1],[1..2147483647]
    3. [case (_, not 0, 0) when b1:] ?DagValueTest(t3 == ConstantValueDefault(0: Int32)) WHEN[b1]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    8. [case (_, _, 0):] ?DagValueTest(t3 == ConstantValueDefault(0: Int32))
    Test: ?DagValueTest(t3 == ConstantValueDefault(0: Int32))
    TrueBranch: 17
    FalseBranch: 20
*State 17 REMAINING t3:[0..0] t1:[-2147483648..-1],[1..2147483647] t2:[-2147483648..-1],[1..2147483647]
    3. [case (_, not 0, 0) when b1:] MATCH WHEN[b1]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    8. [case (_, _, 0):] MATCH
    FalseBranch: 18
*State 18 REMAINING t3:[0..0] t1:[-2147483648..-1],[1..2147483647] t2:[-2147483648..-1],[1..2147483647]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    8. [case (_, _, 0):] MATCH
    FalseBranch: 19
*State 19 REMAINING t3:[0..0] t1:[-2147483648..-1],[1..2147483647] t2:[-2147483648..-1],[1..2147483647]
    8. [case (_, _, 0):] MATCH
*State 20 REMAINING t3:[-2147483648..-1],[1..2147483647] t1:[-2147483648..-1],[1..2147483647] t2:[-2147483648..-1],[1..2147483647]
    5. [case (_, _, _) when b3:] MATCH WHEN[b3]
    FalseBranch: 21
*State 21 FAIL REMAINING t3:[-2147483648..-1],[1..2147483647] t1:[-2147483648..-1],[1..2147483647] t2:[-2147483648..-1],[1..2147483647]
```
</details>